### PR TITLE
Feat/private metadata

### PR DESCRIPTION
**Pull request by NateWilliams2**
- store privateMetadataKey in elections in the db
- global config option for a key to encrypt the privateMetadataKey for each election 
- encrypt & pin metadata for private processes
- fetch & decrypt private metadata for calls authenticated with integrator api token
- also decrypt private metadata for calls with csp signature

### CODE REVIEW

File: urlapi/public.go [hunk 1]
```diff
@@ -211,3 +285,36 @@ func (u *URLAPI) getVoteHandler(msg *bearerstdapi.BearerStandardAPIdata,
...
+	}
+	ecdsaKey, err := ethcrypto.UnmarshalPubkey(rootPub)
+	if err != nil {
+		return fmt.Errorf("could not decode csp public key from election configuration: %w", err)
+	}
+	saltedKey, err := saltedkey.SaltECDSAPubKey(ecdsaKey, processId)
+	if err != nil {
+		return fmt.Errorf("could not salt csp public key: %w", err)
+	}
+	compressedSaltedKey, err := ethereum.CompressPubKey(
```
Comment (p4u) [inline urlapi/public.go:306]: I think it makes more sense to compare the uncompressed pubkeys, it requires less steps

File: urlapi/urlapi.go [hunk 2]
```diff
@@ -62,6 +63,15 @@ func NewURLAPI(router *httprouter.HTTProuter,
 		}
 		log.Infof("global entity encryption key: %s", key)
 	}
+	if len(cfg.GlobalMetaKey) > 0 {
+		key, err := hex.DecodeString(cfg.GlobalMetaKey)
+		if err != nil {
+			log.Errorf("could not decode global metadata key: %v", err)
```
Comment (p4u) [inline urlapi/urlapi.go:69]: Error will continue the execution. I think this should be a Fatal()

File: urlapi/urlapi.go [hunk 3]
```diff
@@ -62,6 +63,15 @@ func NewURLAPI(router *httprouter.HTTProuter,
...
+		} else {
```
Comment (p4u) [inline urlapi/urlapi.go:70]: Also if fatal, there is no need for this else

File: vocclient/vocclient.go [hunk 4]
```diff
@@ -307,6 +332,29 @@ func (c *Client) FetchProcessMetadata(URI string) (*types.ProcessMetadata, error
...
+	}
+	decrypted, ok := apiUtil.DecryptSymmetric(file.Payload, metadataPrivKey)
+	if !ok {
+		return nil, fmt.Errorf("could not decrypt private metadata")
+	}
+	var process types.ProcessMetadata
+	if err = json.Unmarshal(decrypted, &process); err != nil {
+		return nil, err
+	}
+	return &process, nil
```
Comment (p4u) [inline vocclient/vocclient.go:355]: `return &process, json.Unmarshal(decrypted, &process)`

File: vocclient/vocclient.go [hunk 5]
```diff
@@ -275,6 +276,30 @@ func (c *Client) SetProcessMetadata(meta types.ProcessMetadata,
...
+	metaBytes, err = json.Marshal(types.RawFile{Payload: encryptedMeta, Version: "1.0"})
+	if err != nil {
+		return "", fmt.Errorf("could not marshal encrypted bytes: %v", err)
+	}
+	metaURI, err := c.AddFile(metaBytes, "ipfs",
+		fmt.Sprintf("%X process metadata (encrypted)", processId))
+	if err != nil {
+		return "", fmt.Errorf("could not post metadata to ipfs: %v", err)
+	}
+	return metaURI, nil
```
Comment (p4u) [inline vocclient/vocclient.go:300]:
```
return c.AddFile(metaBytes, "ipfs",
		fmt.Sprintf("%X process metadata (encrypted)", processId))
```

File: types/api.go [hunk 6]
```diff
@@ -109,6 +109,11 @@ type ProcessMedia struct {
 	StreamURI string `json:"streamUri,omitempty"`
 }
 
+type RawFile struct {
```
Comment (p4u) [inline types/api.go:112]: Please, add comments to all exported types and methods :1st_place_medal:

File: urlapi/entity.go [hunk 7]
```diff
@@ -449,17 +443,40 @@ func (u *URLAPI) createProcessHandler(msg *bearerstdapi.BearerStandardAPIdata,
...
-	if err != nil {
-		return fmt.Errorf("could not set process metadata: %w", err)
+	var metaUri string
+	var metaPrivKeyBytes []byte
+	var encryptedMetaKey []byte
+	// If election is confidential, generate a private metadata key and encrypt it.
+	// store this key with the election
+	if req.Confidential {
+		metadataPrivKey := dvoteutil.RandomHex(32)
+		if metaPrivKeyBytes, err = hex.DecodeString(metadataPrivKey); err != nil {
```
Comment (p4u) [inline urlapi/entity.go:453]: `if metaPrivKeyBytes, err = hex.DecodeString(dvoteutil.RandomHex(32)); err != nil {`

File: urlapi/entity.go [hunk 8]
```diff
@@ -449,17 +443,40 @@ func (u *URLAPI) createProcessHandler(msg *bearerstdapi.BearerStandardAPIdata,
 		CostExponent:      1,
 	}
 
-	metaUri, err := u.vocClient.SetProcessMetadata(metadata, processID)
-	if err != nil {
-		return fmt.Errorf("could not set process metadata: %w", err)
+	var metaUri string
+	var metaPrivKeyBytes []byte
+	var encryptedMetaKey []byte
```
Comment (p4u) [inline urlapi/entity.go:448]: Is this var needed? why not using `metaPrivKeyBytes`?

File: urlapi/entity.go [hunk 9]
```diff
@@ -449,17 +443,40 @@ func (u *URLAPI) createProcessHandler(msg *bearerstdapi.BearerStandardAPIdata,
...
+				metaPrivKeyBytes, u.globalMetadataKey); err != nil {
+				return fmt.Errorf("could not encrypt metadata private key: %w", err)
+			}
+		}
+
+		if metaUri, err = u.vocClient.SetProcessMetadataConfidential(
+			metadata, metaPrivKeyBytes, processID); err != nil {
+			return fmt.Errorf("could not set confidential process metadata: %w", err)
+		}
+	} else {
```
Comment (p4u) [inline urlapi/entity.go:469]: please, add also comments to the else

File: urlapi/entity.go [hunk 10]
```diff
@@ -536,16 +565,31 @@ func (u *URLAPI) getProcessHandler(
 	}
 
 	// Fetch metadata
-	processMetadata, err := u.vocClient.FetchProcessMetadata(vochainProcess.Metadata)
-	if err != nil {
-		return fmt.Errorf("could not get process metadata: %w", err)
+	var processMetadata *types.ProcessMetadata
+	if dbElection.Confidential {
```
Comment (p4u) [inline urlapi/entity.go:569]: When if/else are so big, I think we should add comments to explain what's happening inside the if statement

File: urlapi/public.go [hunk 11]
```diff
@@ -142,7 +156,67 @@ func (u *URLAPI) getProcessInfoPublicHandler(msg *bearerstdapi.BearerStandardAPI
...
+		}
+	}
+
+	dbElection, err := u.db.GetElectionPrivate(vochainProcess.EntityID, processId)
+	if err != nil {
+		return fmt.Errorf("could not fetch election %x from db: %w", processId, err)
+	}
+
+	err = verifyCspSharedSignature(processId, cspSignature, vochainProcess.CensusRoot)
+	if err != nil {
```
Comment (p4u) [inline urlapi/public.go:188]: `if err := verifyCspSharedSignature(processId, cspSignature, vochainProcess.CensusRoot)`

File: urlapi/public.go [hunk 12]
```diff
@@ -142,7 +156,67 @@ func (u *URLAPI) getProcessInfoPublicHandler(msg *bearerstdapi.BearerStandardAPI
...
+		return fmt.Errorf("shared key not valid to decrypt process %x: %w", processId, err)
+	}
+
+	var processMetadata *types.ProcessMetadata
+	if dbElection.Confidential {
+		metaKey := dbElection.MetadataPrivKey
```
Comment (p4u) [inline urlapi/public.go:194]: Looks like I'm reading the same code several times. Maybe abstract this into a function?

**Review by p4u — CHANGES_REQUESTED** [review]

File: types/types.go [hunk 13]
```diff
@@ -43,20 +43,20 @@ type Organization struct {
...
-	Title            string        `json:"title" db:"title"`
-	CensusID         uuid.NullUUID `json:"censusId" db:"census_id"`
-	StartDate        time.Time     `json:"startDate" db:"start_date"`
-	EndDate          time.Time     `json:"endDate" db:"end_date"`
-	StartBlock       int           `json:"startBlock" db:"start_block"`
-	EndBlock         int           `json:"endBlock" db:"end_block"`
-	Confidential     bool          `json:"confidential" db:"confidential"`
-	HiddenResults    bool          `json:"hiddenResults" db:"hidden_results"`
-	MetadataPrivKey  []byte        `json:"metadataPrivKey" db:"metadata_priv_key"`
+	CreatedUpdated   `json:"created_updated,omitempty"`
```
Comment (emmdim) [inline types/types.go:46]: @NateWilliams2  This is incorrect. Created updated is defined earlier and imported here

File: vocclient/vocclient.go [hunk 14]
```diff
@@ -284,6 +285,27 @@ func (c *Client) SetProcessMetadata(meta types.ProcessMetadata,
...
 
+// SetProcessMetadataConfidential encrypts with metadataPrivKey and then pins
+//  the given process metadata to IPFS and returns its URI
+func (c *Client) SetProcessMetadataConfidential(meta types.ProcessMetadata, metadataPrivKey,
+	processId []byte) (string, error) {
+	metaBytes, err := json.Marshal(meta)
+	if err != nil {
+		return "", fmt.Errorf("could not marshal process metadata: %v", err)
+	}
+	log.Debugf("encrypting meta with %x", metadataPrivKey)
```
Comment (p4u) [inline vocclient/vocclient.go:296]: Lets not print the private key in the logs if possible

**Review by p4u — APPROVED** [review]

File: database/pgsql/elections.go [hunk 15]
```diff
@@ -59,9 +60,17 @@ func (d *Database) GetElectionPublic(organizationEthAddress, processID []byte) (
...
+	selectIntegrator := `SELECT metadata_priv_key, title, census_id, start_date, end_date, start_block, end_block, confidential, hidden_results, integrator_api_key
+						FROM elections WHERE organization_eth_address=$1 AND process_id=$2`
+	row := d.db.QueryRowx(selectIntegrator, organizationEthAddress, processID)
+	return &election, row.StructScan(&election)
+}
+
 func (d *Database) GetElection(integratorAPIKey, orgEthAddress, processID []byte) (*types.Election, error) {
 	var election types.Election
-	selectIntegrator := `SELECT title, census_id, start_date, end_date, start_block, end_block, confidential, hidden_results, 
+	selectIntegrator := `SELECT metadata_priv_key, title, census_id, start_date, end_date, start_block, end_block, confidential, hidden_results,
```
Comment (emmdim) [inline database/pgsql/elections.go:73]: @NateWilliams2  why this is added here since it is returned also in the `getElectionPrivate`?

**Review by emmdim — APPROVED** [review]

**Pull request merged**